crates/wasmtime/src/component/types.rs

```diff
@@ -0,0 +1,1025 @@
+                size: 4,
+                alignment: 4,
+            },
+
+            Type::S64 | Type::U64 | Type::Float64 => SizeAndAlignment {
+                size: 8,
+                alignment: 8,
+            },
+
+            Type::String | Type::List(_) => SizeAndAlignment {
+                size: 8,
+                alignment: 4,
+            },
+
+            Type::Record(handle) => {
+                record_size_and_alignment(handle.fields().map(|field| field.ty))
+            }
+
+            Type::Tuple(handle) => record_size_and_alignment(handle.types()),
+
+            Type::Variant(handle) => variant_size_and_alignment(handle.cases().map(|case| case.ty)),
+
+            Type::Enum(handle) => variant_size_and_alignment(handle.names().map(|_| Type::Unit)),
+
+            Type::Union(handle) => variant_size_and_alignment(handle.types()),
+
+            Type::Option(handle) => {
+                variant_size_and_alignment([Type::Unit, handle.ty()].into_iter())
+            }
+
+            Type::Expected(handle) => {
+                variant_size_and_alignment([handle.ok(), handle.err()].into_iter())
+            }
+
+            Type::Flags(handle) => match FlagsSize::from_count(handle.names().len()) {
+                FlagsSize::Size1 => SizeAndAlignment {
+                    size: 1,
+                    alignment: 1,
+                },
+                FlagsSize::Size2 => SizeAndAlignment {
+                    size: 2,
+                    alignment: 2,
+                },
+                FlagsSize::Size4Plus(n) => SizeAndAlignment {
+                    size: n * 4,
+                    alignment: 4,
+                },
+            },
+        }
+    }
+
+    /// Calculate the aligned offset of a field of this type, updating `offset` to point to just after that field.
+    pub(crate) fn next_field(&self, offset: &mut usize) -> usize {
+        let SizeAndAlignment { size, alignment } = self.size_and_alignment();
+        *offset = func::align_to(*offset, alignment);
+        let result = *offset;
+        *offset += size;
+        result
+    }
+}
+
+fn load_record(
+    types: impl Iterator<Item = Type>,
+    mem: &Memory,
+    bytes: &[u8],
+) -> Result<Box<[Val]>> {
+    let mut offset = 0;
+    types
+        .map(|ty| {
+            ty.load(
+                mem,
+                &bytes[ty.next_field(&mut offset)..][..ty.size_and_alignment().size],
+            )
+        })
+        .collect()
+}
+
+fn lift_variant<'a>(
+    mut types: impl ExactSizeIterator<Item = Type>,
+    store: &StoreOpaque,
+    options: &Options,
+    src: &mut impl Iterator<Item = &'a ValRaw>,
+) -> Result<(u32, Val)> {
+    let len = types.len();
+    let discriminant = next(src).get_u32();
+    let ty = types
+        .nth(discriminant as usize)
+        .ok_or_else(|| anyhow!("discriminant {} out of range [0..{})", discriminant, len))?;
+    let value = ty.lift(store, options, src)?;
+    Ok((discriminant, value))
+}
```

Comment:
I think that this is going to need special logic to skip extra fields within `src` to ensure that `flatten_count` entries are always popped off that iterator.